src/dos/program_mode.cpp

```diff
@@ -0,0 +1,451 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2024-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "program_mode.h"
+
+#include <algorithm>
+#include <set>
+#include <string>
+
+#include "../ints/int10.h"
+#include "ansi_code_markup.h"
+#include "checks.h"
+#include "math_utils.h"
+#include "program_more_output.h"
+#include "regs.h"
+#include "shell.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+constexpr auto FontWidthCommand8Dot = "8dot";
+constexpr auto FontWidthCommand9Dot = "9dot";
+
+enum class BiosFont : uint8_t { _8x8 = 0x12, _8x14 = 0x11, _8x16 = 0x14 };
+
+static void set_font(const BiosFont font)
+{
+	// Load and activate ROM font
+	reg_ah = 0x11;
+	reg_al = enum_val(font);
+
+	// ROM character table 0
+	reg_bl = 0;
+
+	CALLBACK_RunRealInt(0x10);
+}
+
+enum class ScanLines : uint8_t { _200 = 0, _350 = 1, _400 = 2 };
+
+static void set_scan_lines(const ScanLines scan_lines)
+{
+	// Setting the number of scan lines applies to the *next* video mode
+	// change.
+
+	// Select scan lines for text modes
+	reg_ah = 0x12;
+	reg_bl = 0x30;
+	reg_al = enum_val(scan_lines);
+
+	CALLBACK_RunRealInt(0x10);
+}
+
+enum class FontWidth { _8dot, _9dot };
+
+static void set_font_width(const FontWidth width)
+{
+	constexpr auto ClockingModeRegisterIndex = 1;
+
+	IO_Write(0x3c4, ClockingModeRegisterIndex);
+	auto clocking = ClockingModeRegister{IO_Read(0x3c5)};
+
+	// This selects between 8 and 9-dot fonts, which also switches the
+	// horizontal resolution between 640 and 720 pixels in 80-column modes,
+	// and 1056 and 1188 in 132-column modes.
+	clocking.is_eight_dot_mode = (width == FontWidth::_8dot);
+
+	IO_Write(0x3c4, ClockingModeRegisterIndex);
```

Comment:
We have constants for I/O addresses in `inout.h`; it would be nice to continue using this convention.